It is not possible to add MS tile image meta via default_head_blocks.xml

vasilii-b opened the pull request:
### Description (*)
Fixes issue magento/magento2#5023, where you're not able to add a MS tile image with image url from theme.
```xml
<meta name="msapplication-TileImage" content="images/favicons/default/mstile-144x144.png">
````

### Fixed Issues (if relevant)
1. magento/magento2#5023

### Manual testing scenarios (*)
Please check the issue description.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
magento/magento2#5023 - It is not possible to add MS tile image meta via default_head_blocks.xml

magento-engcom-team commented:
Hi @vasilii-b. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

larsroettig commented on lib/internal/Magento/Framework/View/Page/Config.php:
```diff
@@ -539,6 +544,30 @@ public function addBodyClass($className)
+     */
+    public function getMetaAssetUrl($content)
+    {
+        $parsed = parse_url($content);
```
The use of function parse_url() is discouraged

larsroettig replied:
you need to add the second parameter there like PHP_URL_SCHEME

larsroettig commented on lib/internal/Magento/Framework/View/Page/Config.php:
```diff
@@ -539,6 +544,30 @@ public function addBodyClass($className)
+    public function getAdditionalMetaAssets()
+    {
+        return $this->additionalMetaAssets;
+    }
```
Empty Line is missing there .

larsroettig commented on lib/internal/Magento/Framework/View/Test/Unit/Page/ConfigTest.php:
```diff
@@ -369,6 +369,20 @@ public function testAddRss()
         $this->assertInstanceOf(\Magento\Framework\View\Page\Config::class, $this->model->addRss($title, $href));
     }
 
+    public function testGetMetaAssetUrlWithUrlInContent()
```
Function comment is missing there

larsroettig commented on lib/internal/Magento/Framework/View/Test/Unit/Page/ConfigTest.php:
```diff
@@ -369,6 +369,20 @@ public function testAddRss()
+        $this->assetRepo->expects($this->never())->method('getUrl');
+        $this->assertEquals('http://test.com/image.png', $this->model->getMetaAssetUrl('http://test.com/image.png'));
+    }
+
```
Function comment is missing there

Commit:
magento/magento2#5023 - It is not possible to add MS tile image meta via default_head_blocks.xml

# Added codestyle fixes

larsroettig commented on lib/internal/Magento/Framework/View/Page/Config.php:
```diff
@@ -539,6 +544,31 @@ public function addBodyClass($className)
+     */
+    public function getMetaAssetUrl($content)
+    {
+        $parsed = parse_url($content);
```
Hi can you add the second parameter, please

vasilii-b replied:
Hi @larsroettig, I've seen you marked as resolved the previous comment and thought its not needed. 
Will add it. Thank you!

Commit:
magento/magento2#5023 - It is not possible to add MS tile image meta via default_head_blocks.xml

# Changed the way how asset url is checked for URL scheme to a more consistent approach

magento-engcom-team commented:
Hi @larsroettig, thank you for the review.
[ENGCOM-4544](https://jira.corp.magento.com/browse/ENGCOM-4544) has been created to process this Pull Request

VasylShvorak commented:
✔️ QA passed

p-bystritsky committed:
ENGCOM-4544: Semantic Version Checker fix.

sivaschenko commented on lib/internal/Magento/Framework/View/Page/Config.php:
```diff
@@ -539,6 +544,31 @@ public function addBodyClass($className)
+     *
+     * @return array
+     */
+    public function getAdditionalMetaAssets()
```
@vasilii-b @larsroettig adding public methods to an @api class is a backward incompatible change that cannot be accepted to patch version.

Can the code be refactored to avoid that?

i.e. :

- introducing private implementation to Config Renderer
- introducing a new class that provides the required public methods

larsroettig replied:
@sivaschenko sorry I don't see the @api annotations there. Introducing a new class should be possible solutions

sivaschenko reviewed (changes requested):
Hi @vasilii-b thanks for the pull request. The PR contains changes that should be refactored before merging to patch version, please see my code review comments for details.

sidolov commented:
@vasilii-b , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

Commit:
Merge remote-tracking branch 'origin/2.3-develop' into 2.3-layout-head-ms-tile-image-meta-tag

* origin/2.3-develop: (3294 commits)
  magento/magento2#23427 Fix static
  magento/magento2#23532 Fix-static-tests
  magento/magento#23500 fix funk tests
  fix validation class for max-words
  MC-17922: Date in filter area doesn't match with Active filters bar
  Fixed coding standards violations
  Updated unit test
  MC-17922: Date in filter area doesn't match with Active filters bar
  Update Framework/Mail::Message to send all emails as MIME, not just HTML emails.
  MC-17922: Date in filter area doesn't match with Active filters bar
  Removed editors from all dependable customer billing address fields
  Correct array type hints
  Fix mass detection issue
  Testing configurable option update based on product and attribute_id
  move breakpoint by -1px to make nav work correctly at viweport width of 768px
  MC-17275: Admin Down after changing store domain/cookie
  Setting right shipper street1 in addressline1
  MC-17275: Admin Down after changing store domain/cookie
  Replaced urldecode with rawurldecode that not converting + into space
  Fixed issue #23383
  ...

Commit:
magento/magento:5023 - It is not possible to add MS tile image meta via default_head_blocks.xml

# Introduced a new class that provides URL for MS tile image during page config metadata generation

vasilii-b commented:
Hi @sivaschenko, @larsroettig,
Could you guys please check if added changes are good from "Backward compatible" perspective or any other issues are found?
Thank you!

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-5635](https://jira.corp.magento.com/browse/ENGCOM-5635) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Delta commented:
:heavy_check_mark: QA passed

engcom-Foxtrot commented:
@magento run all tests

engcom-Foxtrot committed:
magento/magento2#21798: Static test fix.
